lib/shrine/plugins/derivates.rb

```diff
@@ -0,0 +1,151 @@
+# frozen_string_literal: true
+
+class Shrine
+  module Plugins
+    module Derivates
+      def self.load_dependencies(uploader, *)
+        uploader.plugin :default_url
+      end
+
+      def self.configure(uploader, opts = {})
+        uploader.opts[:derivates_attribute] = opts.fetch(:data_attribute, :separate) # values are separate or original
+        uploader.opts[:version_fallbacks] = opts.fetch(:fallbacks, uploader.opts.fetch(:version_fallbacks, {}))
+        uploader.opts[:versions_fallback_to_original] = opts.fetch(:fallback_to_original, uploader.opts.fetch(:versions_fallback_to_original, true))
+
+        unless [:separate, :original].include? uploader.opts[:derivates_attribute]
+          raise Error, "`#{uploader.opts[:derivates_attribute] || "nil"}` is an invalid " \
+            "option for `data_attribute` in Derivates Plugin. Valid options are: " \
+            "[:separate, :original]"
+        end
+      end
+
+      module ClassMethods
+        def version_fallbacks
+          opts[:version_fallbacks]
+        end
+      end
+
+      module AttachmentMethods
+        def initialize(name, **options)
+          super
+
+          module_eval <<-RUBY, __FILE__, __LINE__ + 1
+            def #{name}_derivates
+              #{name}_attacher.get_derivates
+            end
+          RUBY
+        end
+      end
+
+      module AttacherMethods
+        def changed?
+          instance_variable_defined?(:@old) || instance_variable_defined?(:@derivates_queue)
+        end
+
+        def finalize
+          super
+          update_derivates
+        end
+
+        def add_derivate(key, io, store = :store, nested_list = nil)
+          @derivates_queue ||= {}
+          if store == :store
+            @derivates_queue.merge! key.to_sym => store!(io)
+          else
+            derivate_attacher = shrine_class::Attacher.new(context[:record], context[:name], cache: @cache.storage_key, store: store)
+            @derivates_queue.merge! key.to_sym => derivate_attacher.store!(io)
+          end
+        end
+
+        def update_derivates
+          if instance_variable_defined?(:@derivates_queue)
+            _set_derivates(@derivates_queue)
+            remove_instance_variable(:@derivates_queue)
+          end
+        end
+
+        def get
+          data = read
+          if data
+            if derivates_attribute == :original
+              data = JSON.parse(data)
+              data.delete "derivates" if data.has_key? "derivates"
+            end
+            uploaded_file(data)
+          end
+        end
```

Comment:
Note that `#read` might not always return a JSON string, when using JSON column with ActiveRecord/Sequel it will return a hash.

For now I would recommend this hack (until we think of a proper solution):

```rb
def get
  file = super
  file.instance_variable_set(:@data, file.data.reject { |key| key == "derivates" })
  file
end
```